pkg/cache/tas_flavor_snapshot.go

```diff
@@ -375,15 +377,35 @@ func (s *TASFlavorSnapshot) Fits(flavorUsage workload.TASFlavorUsage) bool {
 	return true
 }
 
+func findPSA(wl *kueue.Workload, psName kueue.PodSetReference) *kueue.PodSetAssignment {
+	if wl == nil || wl.Status.Admission == nil {
+		return nil
+	}
+	for _, psAssignment := range wl.Status.Admission.PodSetAssignments {
+		if psAssignment.Name == psName {
+			return &psAssignment
+		}
+	}
+	return nil
+}
+
 // FindTopologyAssignmentsForFlavor returns TAS assignment, if possible, for all
 // the TAS requests in the flavor handled by the snapshot.
 // The simulateEmpty parameter allows to look for the assignment under the
 // assumption that all TAS workloads are preempted.
-func (s *TASFlavorSnapshot) FindTopologyAssignmentsForFlavor(flavorTASRequests FlavorTASRequests, simulateEmpty bool) TASAssignmentsResult {
+func (s *TASFlavorSnapshot) FindTopologyAssignmentsForFlavor(flavorTASRequests FlavorTASRequests, simulateEmpty bool, wl *kueue.Workload) TASAssignmentsResult {
```

Comment:
From a high level is the following possible, for easier control flow?

At the start of TAS logic, if we are going to do this replacement logic:
1) Remove failed part of the workload's requests from TAS Cache
2) Filter requests down to exactly what needs to be recomputed (the removed part from ^)
3) Determine any domain constraint from the old assignment
4) Now, call the already existing functions with this small request, and the domain constraint
5) Finally, do your merge logic/update annotation, etc

Is such a thing possible to do at the very start, in some contained function? With my superficial glance, it looks like we're injecting conditional logic down the whole call chain, which I think is bug prone and hard to understand.

Reply:
I see your point but if we followed your suggestion we would need to do some computation twice, e.g.:
- calculating `required`
- calculating `levelIdx`

I'm happy to refactor code to follow your proposal just wanted to note that

Reply:
Added, please see if this aligns with your vision